L1Trigger/TrackFindingTracklet/src/KFin.cc

```diff
@@ -0,0 +1,418 @@
+      auto equal = [layer](Stub* stub) { return stub->valid_ && stub->layer_ == layer; };
+      const auto it = find_if(track->stubs_.begin(), track->stubs_.end(), equal);
+      if (it == track->stubs_.end() || !(*it)->valid_)
+        return FrameStub();
+      Stub* stub = *it;
+      const TTBV r(dataFormats_->format(Variable::r, Process::kfin).ttBV(stub->r_));
+      const TTBV phi(dataFormats_->format(Variable::phi, Process::kfin).ttBV(stub->phi_));
+      const TTBV z(dataFormats_->format(Variable::z, Process::kfin).ttBV(stub->z_));
+      const TTBV dPhi(dataFormats_->format(Variable::dPhi, Process::kfin).ttBV(stub->dPhi_));
+      const TTBV dZ(dataFormats_->format(Variable::dZ, Process::kfin).ttBV(stub->dZ_));
+      return FrameStub(stub->ttStubRef_, Frame("1" + r.str() + phi.str() + z.str() + dPhi.str() + dZ.str()));
+    };
+    auto invalid = [](Track* track) { return track && !track->valid_; };
+    auto acc = [invalid](int& sum, Track* track) { return sum += (invalid(track) ? 1 : 0); };
+    const int offsetTrack = region_ * channelAssignment_->numChannelsTrack();
+    for (int channel = 0; channel < channelAssignment_->numChannelsTrack(); channel++) {
+      const int channelTrack = offsetTrack + channel;
+      const int offsetStub = channelTrack * setup_->numLayers();
+      vector<Track*>& input = input_[channel];
+      // fill lost tracks and stubs without gaps
+      const int lost = accumulate(input.begin(), input.end(), 0, acc);
+      lostTracks[channelTrack].reserve(lost);
+      for (int layer = 0; layer < setup_->numLayers(); layer++)
+        lostStubs[offsetStub + layer].reserve(lost);
+      for (Track* track : input) {
+        if (!track || track->valid_)
+          continue;
+        lostTracks[channelTrack].emplace_back(frameTrack(track));
+        for (int layer = 0; layer < setup_->numLayers(); layer++)
+          lostStubs[offsetStub + layer].emplace_back(frameStub(track, layer));
+      }
+      // fill accepted tracks and stubs with gaps
+      acceptedTracks[channelTrack].reserve(input.size());
+      for (int layer = 0; layer < setup_->numLayers(); layer++)
+        accpetedStubs[offsetStub + layer].reserve(input.size());
+      for (Track* track : input) {
+        if (!track || !track->valid_) {  // fill gap
+          acceptedTracks[channelTrack].emplace_back(FrameTrack());
+          for (int layer = 0; layer < setup_->numLayers(); layer++)
+            accpetedStubs[offsetStub + layer].emplace_back(FrameStub());
+          continue;
+        }
+        acceptedTracks[channelTrack].emplace_back(frameTrack(track));
+        for (int layer = 0; layer < setup_->numLayers(); layer++)
+          accpetedStubs[offsetStub + layer].emplace_back(frameStub(track, layer));
+      }
+    }
+  }
+
+  // basetransformation of val from baseLow into baseHigh using widthMultiplier bit multiplication
+  double KFin::redigi(double val, double baseLow, double baseHigh, int widthMultiplier) const {
```

Comment:
I don't understand what this is doing. 
1) Is there a reason why baseLow & baseHigh are not named baseOld & baseNew? Perhaps you assume the old base has a finer granularity than the new one, and redigi() would fail otherwise?
2) If I test it by setting val=100.5 and baselow=basehigh=widthMultiplier=1, then it returns a value of 150.5. Is this expected? What does it mean?

Reply:
this function does as the comment explain convert a value from one into another base.
1. yes I assume conversion from a lower base to a higher base but I believe the function would also work in other cases.
2. the result when I do the same test with 16 bit multiplication is 100.5 as expected, how many bits did you allow for the multiplication?